src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetLockService.cs

```diff
@@ -44,11 +33,15 @@ public NuGetLockService(JoinableTaskContext joinableTaskContext)
         /// </summary>
         public async Task<T> ExecuteNuGetOperationAsync<T>(Func<Task<T>> action, CancellationToken token)
         {
+            var content = NuGetUIThreadHelper.JoinableTaskFactory.Context;
+            var jtc = content.CreateCollection();
+            var jtf = content.CreateFactory(jtc);
```

Comment:
`JoinableTaskContext.CreateFactory(...)` is not going to fix anything here. CPS doesn't have its own `JoinableTaskContext` instance, which is what you're using here to create a factory, so this change looks like it won't fix the issue. CPS *does* have its own `JoinableTaskFactory`, which you're not using here, and may be what you lost in the refactoring done recently that introduced the regression. 
Also in this change, you're creating a new `JoinableTaskCollection` for every single request, which defeats the point of collecting multiple entries and having them join each other to avoid deadlocks. Before this change you shared the collection across multiple invocations as a field, which is the way this should be done.

To mitigate deadlocks with CPS's lock, you should use their JTF specialized instance. But to mitigate your own lock's deadlocks, you need your own JoinableTaskCollection. I think you can achieve both of these by actually creating two nested JTF.RunAsync calls, one using CPS and the other using your JTF instance.

Do you have any tests for this? I'm guessing not, since they would have caught the regression with CPS or the regression that I believe you've introduced in this change. I would recommend prioritizing adding unit tests for this to verify that deadlocks here are never introduced.

Reply:
ohhk I see... I'll change accordingly. About the test, then I wasn't able to repro it locally so I'm not sure of exact repro steps here. For most customers, it happens as part of install/ update operations on SDK based projects. We already have some tests to cover those scenarios but those didn't fail.